bridge-impl/src/main/java/com/liferay/faces/bridge/context/url/internal/BridgeURLBase.java

```diff
@@ -0,0 +1,714 @@
+/**
+ * Copyright (c) 2000-2016 Liferay, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.liferay.faces.bridge.context.url.internal;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.net.MalformedURLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.faces.context.ExternalContext;
+import javax.faces.context.FacesContext;
+import javax.portlet.BaseURL;
+import javax.portlet.MimeResponse;
+import javax.portlet.PortletURL;
+import javax.portlet.ResourceURL;
+import javax.portlet.faces.Bridge;
+import javax.portlet.faces.BridgeUtil;
+
+import com.liferay.faces.bridge.BridgeFactoryFinder;
+import com.liferay.faces.bridge.PortletModeValidator;
+import com.liferay.faces.bridge.PortletModeValidatorFactory;
+import com.liferay.faces.bridge.WindowStateValidator;
+import com.liferay.faces.bridge.WindowStateValidatorFactory;
+import com.liferay.faces.bridge.config.BridgeConfig;
+import com.liferay.faces.bridge.config.internal.BridgeConfigAttributeMap;
+import com.liferay.faces.bridge.context.url.BridgeURI;
+import com.liferay.faces.bridge.context.url.BridgeURL;
+import com.liferay.faces.util.config.ConfiguredServletMapping;
+import com.liferay.faces.util.logging.Logger;
+import com.liferay.faces.util.logging.LoggerFactory;
+
+
+/**
+ * @author  Neil Griffin
+ */
+public abstract class BridgeURLBase implements BridgeURL {
+
+	// Logger
+	private static final Logger logger = LoggerFactory.getLogger(BridgeURLBase.class);
+
+	// Private Data Members
+	private BridgeURI bridgeURI;
+	private List<ConfiguredServletMapping> configuredFacesServletMappings;
+	private String contextPath;
+	private String facesViewTarget;
+	private String namespace;
+	private Map<String, String[]> parameterMap;
+	private String viewId;
+	private String viewIdParameterName;
+
+	public BridgeURLBase(BridgeURI bridgeURI, String contextPath, String namespace, String viewId,
```

Comment:
@ngriffin7a, in my refactor, I've greatly changed this class from [the original](https://github.com/liferay/liferay-faces-bridge-api/pull/7/files#diff-aa0af658d0ff29f5ceade16735f00d61L54). I also made every method (except `toString()`) `final`, `static`, or `private`. I would really like to keep those modifiers even though that might not be the way we usually do things. The reason for adding those modifiers was to indicate to the developer (the future internal Liferay Faces dev) that although `BridgeURLBase` is designed to be extended, the only method that extending classes should override is `toString()`. All other methods should only be called, not overridden. This greatly reduces the developer's options and makes the purpose of the class clearer.